File: Unit-Tests/LUKeychainServicesSpec.m
@import LUKeychainAccess;
#import "Kiwi.h"

SPEC_BEGIN(LUKeychainServicesSpec)

describe(@"LUKeychainServices", ^{
  __block LUKeychainServices *keychainServices;

  beforeEach(^{
    keychainServices = [LUKeychainServices keychainServices];

    [keychainServices deleteAllItemsWithError:nil];
  });

  NSString *key = @"key1", *anotherKey = @"key2";
  NSData *data = [@"data" dataUsingEncoding:NSUTF8StringEncoding], *anotherData = [@"data2" dataUsingEncoding:NSUTF8StringEncoding];

  describe(@"modifying and retrieving data", ^{
    beforeEach(^{
      [keychainServices addData:data forKey:key error:nil];
    });

    it(@"adds and retrieves data from the Keychain", ^{
      [[[keychainServices dataForKey:key error:nil] should] equal:data];
    });

    it(@"updates data from the keychain", ^{
      [keychainServices updateData:anotherData forKey:key error:nil];

      [[[keychainServices dataForKey:key error:nil] should] equal:anotherData];
    });
  });

  describe(@"deleting items", ^{
    beforeEach(^{
      [keychainServices addData:data forKey:key error:nil];
      [keychainServices addData:anotherData forKey:anotherKey error:nil];
    });

    it(@"deletes items with a given key", ^{
      [keychainServices deleteItemWithKey:key error:nil];

      NSError *error;
      [[[keychainServices dataForKey:key error:&error] should] beNil];
      [[theValue(error.code) should] equal:theValue(errSecItemNotFound)];

      [[[keychainServices dataForKey:anotherKey error:nil] should] equal:anotherData];
    });

    it(@"deletes all items in the keychain", ^{
      [keychainServices deleteAllItemsWithError:nil];

      NSError *error;
      [[[keychainServices dataForKey:key error:&error] should] beNil];
      [[theValue(error.code) should] equal:theValue(errSecItemNotFound)];

      error = nil;
      [[[keychainServices dataForKey:anotherKey error:&error] should] beNil];
      [[theValue(error.code) should] equal:theValue(errSecItemNotFound)];
    });
  });
});

SPEC_END

File: Swift UnitTests/LUKeychainServicesSpec.swift
//
//  LUKeychainServicesSpec.swift
//  Swift UnitTests
//
//  Copyright © 2022 GrubHub. All rights reserved.
//

import XCTest
import Nimble
import Quick
@testable import LUKeychainAccess

final class LUKeychainServicesSpec: QuickSpec {
  override func spec() {
    var services: LUKeychainServices!
    var error: NSError?
    var error2: NSError?
    let key = "key1"
    let key2 = "key2"
    let data =  "data".data(using: .utf8)!
    let data2 = "data2".data(using: .utf8)!
    
    beforeEach {
      services = LUKeychainServices()
      services.deleteAllItems(error: nil)
      error = nil
      error2 = nil
    }
    
    describe("modifying and retrieving data") {
      beforeEach {
        services.add(data, for: key, error: nil)
      }
      
      it("adds and retrieves data from the keychain") {
        let result = services.data(for: key, error: &error)
        expect(result).to(equal(data))
        expect(error).to(beNil())
      }
      
      it("updates data from the keychain") {
        services.update(data2, for: key, error: nil)
        let result = services.data(for: key, error: nil)
        expect(result).to(equal(data2))
        expect(error).to(beNil())
      }
    }
    
    describe("deleting items") {
      beforeEach {
        services.add(data, for: key, error: nil)
        services.add(data2, for: key2, error: nil)
      }
      
      it("deletes only the item with a given key") {
        services.deleteItem(for: key, error: &error)
        expect(error).to(beNil())
        expect(services.data(for: key, error: &error2)).to(beNil())
        expect(error2).toNot(beNil())
        expect(Int(error2!.code)).to(equal(Int(errSecItemNotFound)))
        expect(services.data(for: key2, error: nil)).to(equal(data2))
      }
      
      it("deletes all items") {
        services.deleteAllItems(error: &error)
        expect(error).to(beNil())
        expect(services.data(for: key, error: &error)).to(beNil())
        expect(error).toNot(beNil())
        expect(Int(error!.code)).to(equal(Int(errSecItemNotFound)))
        expect(services.data(for: key2, error: &error2)).to(beNil())
        expect(error2).toNot(beNil())
        expect(Int(error2!.code)).to(equal(Int(errSecItemNotFound)))
      }
    }
  }
}
